engines/director/lingo/lingo-bytecode.cpp

```diff
@@ -0,0 +1,311 @@
+	stream.readUint16();
+	stream.readUint16();
+	stream.readUint16();
+	uint16 constsBase = stream.readUint16();
+
+	// preload all the constants!
+	// these are stored as a reference table of 6 byte entries, followed by a storage area.
+
+	// copy the storage area first.
+	uint32 constsStoreOffset = constsOffset + 6 * constsCount;
+	uint32 constsStoreSize = stream.size() - constsStoreOffset;
+	stream.seek(constsStoreOffset);
+	byte *constsStore = (byte *)malloc(constsStoreSize);
+	stream.read(constsStore, constsStoreSize);
+
+	Common::Array<Datum> constsData;
+
+	// read each entry in the reference table.
+	stream.seek(constsOffset);
+	for (uint16 i=0; i < constsCount; i++) {
+		Datum constant;
+		uint16 constType = stream.readUint16();
+		uint32 value = stream.readUint32();
+		switch (constType) {
+			case 1: // String type
+				{
+					constant.type = STRING;
+					constant.u.s = new Common::String();
+					uint32 pointer = value;
+					while (pointer < constsStoreSize) {
+						if (constsStore[pointer] == '\r') {
+							constant.u.s += '\n';
+						} else if (constsStore[pointer] == '\0') {
+							break;
+						} else {
+							constant.u.s += constsStore[pointer];
+						}
+						pointer += 1;
+					}
+					if (pointer >= constsStoreSize) {
+						warning("Constant string has no null terminator");
+						break;
+					}
+				}
+				break;
+			case 4: // Integer type
+				constant.type = INT;
+				constant.u.i = (int)value;
+				break;
+			case 9:  // Float type
+				{
+					constant.type = FLOAT;
+					if (value < constsStoreOffset) {
+						warning("Constant float start offset is out of bounds");
+						break;
+					} else if (value + 4 > constsStoreSize) {
+						warning("Constant float end offset is out of bounds");
+						break;
+					}
+					constant.u.f = *(float *)(constsStore + value);
+				}
+				break;
+			default:
+				warning("Unknown constant type %d", type);
+				break;
+		}
+
+		constsData.push_back(constant);
+	}
+	free(constsStore);
+
+	// parse each function!
+	// these are stored as a code storage area, followed by a reference table of 42 byte entries.
+
+	// copy the storage area first.
+	uint32 codeStoreSize = functionsOffset - codeStoreOffset;
+	stream.seek(codeStoreOffset);
+	byte *codeStore = (byte *)malloc(codeStoreSize);
+	stream.read(codeStore, codeStoreSize);
+
+	// read each entry in the function table.
+	stream.seek(functionsOffset);
+	for (uint16 i=0; i<functionsCount; i++) {
+		_currentScriptFunction = i;
+		_currentScriptContext->functions.push_back(new ScriptData);
+		_currentScript = _currentScriptContext->functions[_currentScriptFunction];
+
+		uint16 nameIndex = stream.readUint16();
+		stream.readUint16();
+		uint32 length = stream.readUint32();
+		uint32 startOffset = stream.readUint32();
+		uint16 argCount = stream.readUint16();
+		uint32 argOffset = stream.readUint32();
+		uint16 varCount = stream.readUint16();
+		uint32 varNamesOffset = stream.readUint32();
+		stream.readUint16();
```

Comment:
Would be beneficial to print these out on debug level 8+, so we could potentially figure out their meaning.

Reply:
Done

Reply:
I think there is a misunderstanding. What I meant here is to do the following:
```unk1 = stream.readUint16();
unk2 = stream.readUint16();
...
debug(8, "funcitionentry: nameIndex=%d, unk1=%d, length=%d ...", nameIndex, unk1, length, unk2, unk3...);